docs/device-api.md

```diff
@@ -0,0 +1,106 @@
+# Device-API Functionality
+
+The Device-API following functionality matrix indicates which devices are
+required for an API call to be usable.
+
+**O** (Optional): The device (column) **is not required** for a Firecracker
+microVM to run. If the device (column) is omitted from a kernel build the API
+element (row) is **TBA**.
```

Comment:
The request for an API endpoint which lacks the needed Firecracker component for that API to complete returns a response with an error.

Reply:
OK. However, we have not used the word component on this page so that statement will need to be reworded, or the definition of 'component' added somewhere herein.

Reply:
Not sure what to place here in the stead of **TBA**.

Reply:
Please consider the following suggestion. WDYT?
```suggestion
element request (row) returns with an HTTP response containing a corresponding error message.
```

Reply:
Just another thing which I realized, since the table refers to the device requirement for APIs to succeed, is that `/mmds` and `/mmds/config` does not have any hard requirement on the existence of a configured network device. A network device is relevant only for MMDS operation, but not for its configuration. With this in mind, I think it makes sense to change the `virtio-net` column value corresponding to both `/mmds` and `/mmds/config` from `**R**` to `**O**`.

Reply:
I've not used MMDS in anger.  My instinct is that configuring something and not using it sounds like an unusual corner case?
Is there a mainstream use case where MMDS would be configured for a uVM that does not have a network device, but that MMDS configuration is then used on a uVM with a network device?

Reply:
> request (row) returns with an HTTP response containing a corresponding error message.

Is there a link to the error messages we can use, or should that be documented here?

Reply:
> Is there a mainstream use case where MMDS would be configured for a uVM that does not have a network device, but that MMDS configuration is then used on a uVM with a network device?
I am not aware of any mainstream case as the one you've described.

>  My instinct is that configuring something and not using it sounds like an unusual corner case.
I agree. The misalignment which might cause customer confusion comes from the fact that the MMDS configuration requirements does not correspond with MMDS operation requirements. I can not say this is a required design choice or simply an arbitrary happening due to unclear requirements. My understanding is that it is not a blocker of any sort for Firecracker usage and configuration since we haven't had any issues with it so far. I also believe that with the advent of the Device-API matrix it will be easier to think for the future in terms of API refinement in this area. Feel free to open any issue on this topic.

Reply:
I do not think we need to handle this specific occurrence of TBA as in what errors would an end user get if the optional requirements are not satisfied. Here we should say that the API endpoint call will succeed if the requirement is not satisfied. I would add an expectancy in terms of returned HTTP responses only for the `Required` section.

> Is that a 404 response?

We do not have yet an unified interface for expected responses if users do not respect mandatory requirements of an API endpoint. At the moment, we can work backwards from two API endpoints: `drives/{id}` and `network-interfaces/{id}`. Only these two have a hard mandatory requirement on the existence of a configured virtio-block/virtio-net device with the same `id` used by the API call.

* `drives/{id}` results in 400 - BadRequest - HTTP response if the block device identified by `id` was not previously configured.
* `network-interfaces/{id}` API call will act as a no-op if the network device identified by `id` was not previously configured.

> Is there a link to the error messages we can use, or should that be documented here?

We do not have a documentation for the error messages returned in response to `drives/{id}` or `network-interfaces/{id}`. I would suggest you to mention the returned errors response code for `drives/{id}` and mention also that `network-interfaces/{id}` is considered as no-op in case of error. This is not unified across devices. Feel free to open an issue for this as well.

Reply:
Done.  What about `vsock`?

Reply:
@bbros-dev , please see my comment updates above.

> What about vsock?

`vsock` API endpoint call does not require the pre-existance of a configured `vsock` device, so we do not need **R** for virtio-vsock section of the `vsock` API endpoint. If we have it **O** then we do not need to set an expectancy for not satisfying the API endpoint mandatory requirements, since there are not any.

> Also `SendCtlAltDel` when no keyboard is present?

The keyboard device is always present for a Firecracker uVM. On a separate note, this API endpoint is only available on `x86_64`. Firecracker at the moment has arch dependencies for this API endpoint and for `CreateSnapshot`/`LoadSnapshot` as well. I think this is another dependency we want to capture in the document, but with this information the document scope will change from `Device-API matrix` to `Firecracker API dependency matrix`. Please let me know what you think.

Reply:
>  I think this is another dependency we want to capture in the document, but with this information the document scope will change from  Device-API matrix to Firecracker API dependency matrix. Please let me know what you think.

Agree. Perhaps an separate issue to track updating this with ARM64 specifics once that architecture is stabilized?

Reply:
Yes, we can definitely work with another issue for tracking Firecracker API arch dependency dimension. Feel free to open it.

Reply:
Okay, so we are done?